client/authority-discovery/src/worker/addr_cache.rs

```diff
@@ -17,98 +17,93 @@
-use std::collections::HashMap;
 
 use sc_network::PeerId;
 use sp_authority_discovery::AuthorityId;
+use std::collections::{hash_map::Entry, HashMap, HashSet};
 
 /// Cache for [`AuthorityId`] -> [`Vec<Multiaddr>`] and [`PeerId`] -> [`AuthorityId`] mappings.
 pub(super) struct AddrCache {
-	// The addresses found in `authority_id_to_addresses` are guaranteed to always match
-	// the peerids found in `peer_id_to_authority_id`. In other words, these two hashmaps
-	// are similar to a bi-directional map.
-	authority_id_to_addresses: HashMap<AuthorityId, Vec<Multiaddr>>,
-	peer_id_to_authority_id: HashMap<PeerId, AuthorityId>,
+	/// The addresses found in `authority_id_to_addresses` are guaranteed to always match
+	/// the peerids found in `peer_id_to_authority_ids`. In other words, these two hashmaps
+	/// are similar to a bi-directional map.
+	authority_id_to_addresses: HashMap<AuthorityId, HashSet<Multiaddr>>,
+	peer_id_to_authority_ids: HashMap<PeerId, HashSet<AuthorityId>>,
 }
 
 impl AddrCache {
 	pub fn new() -> Self {
 		AddrCache {
 			authority_id_to_addresses: HashMap::new(),
-			peer_id_to_authority_id: HashMap::new(),
+			peer_id_to_authority_ids: HashMap::new(),
 		}
 	}
 
 	/// Inserts the given [`AuthorityId`] and [`Vec<Multiaddr>`] pair for future lookups by
 	/// [`AuthorityId`] or [`PeerId`].
-	pub fn insert(&mut self, authority_id: AuthorityId, mut addresses: Vec<Multiaddr>) {
-		addresses.sort_unstable_by(|a, b| a.as_ref().cmp(b.as_ref()));
-
-		// Insert into `self.peer_id_to_authority_id`.
-		let peer_ids = addresses
-			.iter()
-			.map(|a| peer_id_from_multiaddr(a))
-			.filter_map(|peer_id| peer_id);
-		for peer_id in peer_ids.clone() {
-			let former_auth =
-				match self.peer_id_to_authority_id.insert(peer_id, authority_id.clone()) {
-					Some(a) if a != authority_id => a,
-					_ => continue,
-				};
-
-			// PeerId was associated to a different authority id before.
-			// Remove corresponding authority from `self.authority_id_to_addresses`.
-			let former_auth_addrs = match self.authority_id_to_addresses.get_mut(&former_auth) {
-				Some(a) => a,
-				None => {
-					debug_assert!(false);
-					continue
-				},
-			};
-			former_auth_addrs.retain(|a| peer_id_from_multiaddr(a).map_or(true, |p| p != peer_id));
-		}
-
-		// Insert into `self.authority_id_to_addresses`.
-		for former_addr in self
-			.authority_id_to_addresses
-			.insert(authority_id.clone(), addresses.clone())
-			.unwrap_or_default()
-		{
-			// Must remove from `self.peer_id_to_authority_id` any PeerId formerly associated
-			// to that authority but that can't be found in its new addresses.
-
-			let peer_id = match peer_id_from_multiaddr(&former_addr) {
-				Some(p) => p,
-				None => continue,
-			};
+	pub fn insert(&mut self, authority_id: AuthorityId, addresses: Vec<Multiaddr>) {
+		let addresses = addresses.into_iter().collect::<HashSet<_>>();
+		let peer_ids = addresses_to_peer_ids(&addresses);
```

Comment:
maybe we should ensure the len is 1 here? See the comment above

Reply:
Done

Reply:
I actually meant exactly one

Reply:
Hmm, I mean you could change your peerid and then for some time you would occur with multiple peer ids in the dht?

Even if someone runs two nodes, we can not know this for sure.

Reply:
But if you're putting a new record, won't it override the old one? They are keyed by `AuthorityId`(s). 

> Even if someone runs two nodes, we can not know this for sure.

yeah, I wrote this in the comment above:
> Technically, one could copy an AuthorityId to a validator with different PeerId and change other session keys. But that's a really contrived scenario.

Reply:
> But if you're putting a new record, won't it override the old one? They are keyed by `AuthorityId`(s).

I don't exactly know how our dht works. However, from what I remember from back in the days, where I read about dhts, it is rather hard to make sure that a record is updated everywhere (connections could have changed etc). So, I would not bet that a change propagates that fast. At some point I assume that you are right and the old value should be overwritten, but it probably takes some time.

Maybe @tomaka knows more.

Reply:
Even if there are two records (old and new) floating around, we'll never see a mix of a few records. We either see one or another. And each record should contain multiaddr corresponding to one peer. Otherwise, a warning might be warranted -- this is either an equivocation symptom or misbehavior.

Reply:
Even if the current implementation does not end up with multiple PeerId-s, does it really affect security if the same AuthorityId is used on multiple peers? That would need copying the ephemeral private key of the session from one peer to another, right?

Reply:
An authority that has put multiple PeerIds into its own record can do harm to noone but itself, by not properly connecting to everyone an missing on approval voting rewards. But I think we should detect and report this and not silently ignore.

Reply:
So are you okay if finding multiple PeerId-s for an AuthorityId is detected and logged, but not forbidden?

Reply:
>  But I think we should detect and report this and not silently ignore.

Done